src/query/api/v1/middleware/logging.go

```diff
@@ -50,19 +53,60 @@ func RequestID(iOpts instrument.Options) mux.MiddlewareFunc {
 	}
 }
 
+// LoggingOptions are the options for the logging middleware.
+type LoggingOptions struct {
+	Threshold time.Duration
+	Fields    Fields
+}
+
+// WithNoResponseLogging disables response logging for a route.
+var WithNoResponseLogging = func(opts Options) Options {
+	opts.Logging.Threshold = 0
+	return opts
+}
+
+// Fields returns additional logging fields to add to the response log message.
+type Fields func(r *http.Request, start time.Time) []zap.Field
+
 // ResponseLogging logs the response time if the request took longer than the configured threshold.
-func ResponseLogging(threshold time.Duration, iOpts instrument.Options) mux.MiddlewareFunc {
+func ResponseLogging(opts Options) mux.MiddlewareFunc {
 	return func(base http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			startTime := time.Now()
+			start := opts.Clock.Now()
+			statusCodeTracking := &xhttp.StatusCodeTracker{ResponseWriter: w, TrackError: true}
+			w = statusCodeTracking.WrappedResponseWriter()
 			base.ServeHTTP(w, r)
-			d := time.Since(startTime)
-			if threshold > 0 && d >= threshold {
-				logger := logging.WithContext(r.Context(), iOpts)
-				logger.Info("finished handling request",
+			d := opts.Clock.Now().Sub(start)
+			if opts.Logging.Threshold > 0 && d >= opts.Logging.Threshold {
+				logger := logging.WithContext(r.Context(), opts.InstrumentOpts)
+				fields := []zap.Field{
 					zap.Duration("duration", d),
-					zap.String("url", r.URL.RequestURI()))
+					zap.String("url", r.URL.RequestURI()),
+					zap.Int("status", statusCodeTracking.Status),
+				}
+				if statusCodeTracking.ErrMsg != "" {
+					fields = append(fields, zap.String("error", statusCodeTracking.ErrMsg))
+				}
+				fields = appendM3Headers(r.Header, fields)
+				fields = appendM3Headers(w.Header(), fields)
+				if opts.Logging.Fields != nil {
```

Comment:
this will make it much easier to add custom headers to the log fields in the future